File: client/tws_foundation_client/lib/src/models/options/set_view_options.dart
import 'package:csm_client/csm_client.dart';
import 'package:tws_foundation_client/tws_foundation_client.dart';

///
final class SetViewOptions<TSet extends CSMSetInterface> implements CSMEncodeInterface {
  ///
  final bool retroactive;

  ///
  final int range;

  ///
  final int page;

  ///
  final DateTime? creation;

  ///
  final List<SetViewOrderOptions> orderings;

  ///
  final List<SetViewFilterNodeInterface<TSet>> filters;

  ///
  const SetViewOptions(this.retroactive, this.range, this.page, this.creation, this.orderings, this.filters);

  @override
  JObject encode() {
    return <String, dynamic>{
      'retroactive': retroactive,
      'range': range,
      'page': page,
      'creation': creation,
      'orderings': orderings.map((SetViewOrderOptions i) => i.encode()).toList(),
      'filters': filters.map((SetViewFilterNodeInterface<TSet> i) => i.encode()).toList(),
    };
  }
}

File: server/TWS Foundation/CSM Foundation/Database/Models/Options/SetViewOptions.cs
using CSM_Foundation.Database.Interfaces;

namespace CSM_Foundation.Database.Models.Options;

/// <summary>
///     Defines options to build a <see cref="MigrationView"/> 
///     specifing the behavior to the builder.
/// </summary>
public class SetViewOptions<TSet> 
    where TSet : ISet {
    
    /// <summary>
    ///     On <see langword="true"/> indicate that the builder should consider all the new items added 
    ///     after the <see cref="Creation"/> if it is null then won't consider the limitation will behave
    ///     as this property is <see langword="true"/>
    /// </summary>
    public required bool Retroactive { get; init; }
    
    /// <summary>
    ///     Specifies the amount of items expected per page
    /// </summary>
    public required int Range { get; init; }
    
    /// <summary>
    ///     Specifies the current desired page.
    /// </summary>
    public required int Page { get; init; }
    
    /// <summary>
    ///     Specifies the last time this view was created, this works to limit the new entries 
    ///     on demand by <see cref="Retroactive"/>
    /// </summary>
    public DateTime? Creation { get; init; }
    
    /// <summary>
    ///     Indicates order actions to perform to the current view building.
    ///     <br>
    ///         The ordering actions will be performed by the array received order.
    ///     </br>
    /// </summary>
    public SetViewOrderOptions[] Orderings { get; init; } = [];

    /// <summary>
    /// 
    /// </summary>
    public ISetViewFilterNode<TSet>[] Filters { get; init; } = [];
}
